linkerd/app/inbound/src/lib.rs

```diff
@@ -127,32 +123,88 @@ impl Config {
-                svc::stack(tcp_forward.clone())
-                    .push_map_target(TcpEndpoint::from)
-                    .push(svc::NewUnwrapOr::layer(
-                        svc::Fail::<_, NonOpaqueRefused>::default(),
-                    ))
-                    .push(detect::NewDetectService::timeout(
-                        self.proxy.detect_protocol_timeout,
-                        transport_header::DetectHeader::default(),
-                    )),
-            )
-            .into_inner();
+        // The direct stack handles connections that target the proxy's inbound
+        // port (i.e. without an SO_ORIGINAL_DST setting). This port behaves
+        // differently from the main proxy stack:
+        //
+        // 1. Protocol detection is always performed;
+        // 2. TLS is required;
+        // 3. A transport header is expected. It's not strictly required, as
+        //    gateways may need to accept HTTP requests from older proxy versions.
+        let direct = {
+            let http = svc::stack(http_loopback)
+                .push_on_response(
+                    svc::layers()
+                        .push(svc::FailFast::layer("Gateway", dispatch_timeout))
+                        .push_spawn_buffer(self.proxy.buffer_capacity),
+                )
+                .check_new_service::<Target, http::Request<http::BoxBody>>()
+                .into_inner();
+
+            // If the connection targets the inbound port, try to detect an
+            // opaque transport header and rewrite the target port accordingly.
+            // If there was no opaque transport header, fail the connection with
+            // a ConnectionRefused error.
+            svc::stack(self.http_server(http, &metrics, span_sink.clone(), drain.clone()))
+                // When HTTP detection fails, forward the connection to the
+                // application as an opaque TCP stream.
+                .push(svc::NewUnwrapOr::layer(
+                    tcp_forward
+                        .clone()
+                        .push_request_filter(prevent_loop)
+                        .push_map_target(TcpEndpoint::from)
+                        .into_inner(),
+                ))
+                .push_cache(self.proxy.cache_max_idle_age)
+                // Update the TcpAccept target using a parsed transport-header,
+                // if there was one.
+                //
+                // TODO use the transport header's `name` to inform gateway
+                // routing.
+                .push_map_target(
+                    |(v, (h, mut t)): (
+                        Option<http::Version>,
+                        (Option<TransportHeader>, TcpAccept),
+                    )| {
+                        if let Some(h) = h {
+                            t.target_addr = (t.target_addr.ip(), h.port).into();
+                        }
+                        (v, t)
+                    },
```

Comment:
yeesh.

(i imagine this is going to change soon though, based on the comment?)